bls-signatures/src/signature.rs

```diff
@@ -154,6 +212,224 @@ impl SignatureProjective {
+    /// identifying which signatures are valid.
+    ///
+    /// This function uses a highly optimized parallel binary search approach (divide and conquer)
+    /// to efficiently identify invalid signatures in the batch (O(k log N) pairings).
+    #[cfg(feature = "parallel")]
+    pub fn par_verify_batch_binary_search<
+        P: AsPubkeyProjective + Sync,
+        S: AsSignatureProjective + Sync,
+    >(
+        public_keys: &[&P],
+        signatures: &[&S],
+        message: &[u8],
+        options: &VerificationOptions,
+    ) -> Result<Vec<bool>, BlsError> {
+        if public_keys.len() != signatures.len() {
+            return Err(BlsError::InputLengthMismatch);
+        }
+
+        // Convert all inputs to the internal projective representation *once*
+        // to avoid repeated expensive conversions (e.g., decompression) during recursion.
+        let inputs: Result<Vec<(PubkeyProjective, SignatureProjective)>, BlsError> = public_keys
+            .par_iter()
+            .zip(signatures.par_iter())
+            .map(|(pk, sig)| {
+                // If any conversion fails (e.g., invalid point encoding), the error propagates.
+                let pk_proj = pk.try_as_projective()?;
+                let sig_proj = sig.try_as_projective()?;
+                Ok((pk_proj, sig_proj))
+            })
+            .collect();
+
+        let inputs = inputs?;
+        let n = inputs.len();
+
+        // Initialize results to false (invalid until proven valid).
+        let mut results = alloc::vec![false; n];
+
+        // Recursive identification.
+        Self::recursive_identify(
+            &inputs,
+            message,
+            &mut results,
+            options.aggregation_threshold,
+        );
+
+        Ok(results)
+    }
+
+    /// Helper function to perform the parallelized binary search on pre-validated inputs.
+    #[cfg(feature = "parallel")]
+    #[allow(clippy::arithmetic_side_effects)]
+    fn recursive_identify(
+        inputs: &[(PubkeyProjective, SignatureProjective)],
+        message: &[u8],
+        results: &mut [bool],
+        aggregation_threshold: NonZeroUsize,
+    ) {
+        if inputs.is_empty() {
+            return;
+        }
+
+        // Aggregate the current batch.
+
+        // For small subsets, sequential aggregation is often faster due to Rayon overhead.
+        let (aggregate_pubkey, aggregate_signature) = if inputs.len() < aggregation_threshold.into()
+        {
+            // Sequential aggregation
+            inputs.iter().fold(
+                (
+                    PubkeyProjective::identity(),
+                    SignatureProjective::identity(),
+                ),
+                |(mut pk_acc, mut sig_acc), (pk, sig)| {
+                    pk_acc.0 += &pk.0;
+                    sig_acc.0 += &sig.0;
+                    (pk_acc, sig_acc)
+                },
+            )
+        } else {
+            // Parallel aggregation
```

Comment:
Any performance difference between the two approaches?

Also, because we already have the pubkey and signature aggregation for the whole array, can we do
pubkey_right = pubkey_whole - pubkey_left
To save half of the aggregation?

Reply:
In terms of `pubkey_right = pubkey_whole - pubkey_left`, we can definitely do that as a form of optimization. To support that, we should create a function that takes in an already aggregated public key instead of a vector/slice of public keys.

But since this is an optimization, I think we should complete sigverify with the current API initially and then incrementally add optimization.